bin/metrics-cloudfront.rb

```diff
@@ -0,0 +1,149 @@
+         description: 'Distribution id of Cloudfront',
+         short: '-d DISTRIBUTION_ID',
+         long: '--distribution_id DISTRIBUTION_ID'
+
+  option :scheme,
+         description: 'Metric naming scheme, text to prepend to metric',
+         short: '-s SCHEME',
+         long: '--scheme SCHEME',
+         default: 'aws.cloudfront'
+
+  option :aws_region,
+         short: '-r AWS_REGION',
+         long: '--aws-region REGION',
+         description: 'AWS Region (defaults to us-east-1).',
+         default: 'us-east-1'
+
+  option :metric,
+         description: 'Metric to fetch',
+         short: '-m METRIC',
+         long: '--metric'
+
+  option :end_time,
+         short:       '-t T',
+         long:        '--end-time TIME',
+         default:     Time.now,
+         proc:        proc { |a| Time.parse a },
+         description: 'CloudWatch metric statistics end time'
+
+  option :period,
+         short:       '-p N',
+         long:        '--period SECONDS',
+         default:     60,
+         proc:        proc(&:to_i),
+         description: 'CloudWatch metric statistics period'
+
+  def cloud_watch
+    @cloud_watch ||= Aws::CloudWatch::Client.new
+  end
+
+  def cloud_watch_metric(metric_name, value, distribution_id)
+    cloud_watch.get_metric_statistics(
+      namespace: 'AWS/CloudFront',
+      metric_name: metric_name,
+      dimensions: [
+        {
+          name: 'Region',
+          value: 'Global'
+        },
+        {
+          name: 'DistributionId',
+          value: distribution_id
+        }
+      ],
+      statistics: [value],
+      start_time: config[:end_time] - config[:period],
+      end_time: config[:end_time],
+      period: config[:period]
+    )
+  end
+
+  def distribution_list
+    list_metrics = cloud_watch.list_metrics(
+      namespace: 'AWS/CloudFront'
+    ).metrics
+
+    list_metrics = list_metrics.select { |e| e.metric_name == config[:metric] } unless config[:metric].nil?
+
+    list_metrics.reduce(Set.new) do |result, item|
+      result << item.dimensions.find { |element| element.name == 'DistributionId' }.value
+    end
+  end
+
+  def print_statistics(distribution_id, statistics)
+    statistics.each do |key, static|
+      r = cloud_watch_metric(key, static, distribution_id)
+      keys = [config[:scheme]]
+      keys.concat [distribution_id, key, static]
+      output(keys.join('.'), r[:datapoints].first[static.downcase]) unless r[:datapoints].first.nil?
+    end
+  end
+
+  def run
+    statistic = {
+      'Requests' => 'Sum',
+      'BytesDownloaded' => 'Sum',
+      'BytesUploaded' => 'Sum',
+      'TotalErrorRate' => 'Average',
+      '4xxErrorRate' => 'Average',
+      '5xxErrorRate' => 'Average'
+    }
+
+    unless config[:metric].nil?
+      statistic.select! { |key, _| key == config[:metric] }
+    end
+
+    if statistic.empty?
+      critical 'metric can be (Requests, BytesDownloaded, BytesUploaded, TotalErrorRate, 4xxErrorRate, 5xxErrorRate)'
```

Comment:
rather than giving an error message why not just make it required and provide those values as input list?

example: https://github.com/sensu-plugins/sensu-plugins-redis/blob/3.0.0/lib/redis_client_options.rb#L75

Reply:
@majormoses I didn't know about this possibility.
fixed